core/src/main/kotlin/xtdb/util/TemporalBounds.kt

```diff
@@ -4,34 +4,66 @@ import java.time.Instant
-        fun lte(operand: Long) = apply { upper = min(operand, upper) }
-        fun gt(operand: Long) = apply { lower = max(operand + 1, lower) }
-        fun gte(operand: Long) = apply { lower = max(operand, lower) }
-
-        fun inRange(operand: Long) = operand in lower..upper
+    fun expand(operand: Long) = apply {
+        lower = min(lower, operand)
+        upper = max(upper, operand)
+    }
 
-        override fun toString(): String {
-            val l = Instant.ofEpochMilli(lower / 1000)
-            val u = Instant.ofEpochMilli(upper / 1000)
-            return "TemporalColumn(lower=$l, upper=$u)"
-        }
+    override fun toString(): String {
+        val l = Instant.ofEpochMilli(lower / 1000)
+        val u = Instant.ofEpochMilli(upper / 1000)
+        return "TemporalColumn(lower=$l, upper=$u)"
     }
+}
+
+data class TemporalBounds(
+    val validFrom: TemporalColumn = TemporalColumn(),
+    val validTo: TemporalColumn = TemporalColumn(),
+    val systemFrom: TemporalColumn = TemporalColumn(),
+    val systemTo: TemporalColumn = TemporalColumn()) {
 
     fun inRange(validFrom: Long, validTo: Long, systemFrom: Long, systemTo: Long) =
         (this.validFrom.inRange(validFrom)
             && this.validTo.inRange(validTo)
             && this.systemFrom.inRange(systemFrom)
             && this.systemTo.inRange(systemTo))
+
+    // intersects - There exists at least one interval (rectangle) that is common to both bounds
+
+    fun intersectsValidTime(other: TemporalBounds) = (validFrom.intersects(other.validFrom) && validTo.intersects(other.validTo))
+    fun intersectsSystemTime (other: TemporalBounds) = (systemFrom.intersects(other.systemFrom) && systemTo.intersects(other.systemTo))
+    fun intersects(other: TemporalBounds) = intersectsValidTime(other) && intersectsSystemTime(other)
+
+    // overlaps - There exists two intervals (rectangles), one from each bound, that overlap
+
+    fun overlapsValidTime(other: TemporalBounds) = (validFrom.intersects(other.validFrom) || validTo.intersects(other.validTo))
+    fun overlapsSystemTime(other: TemporalBounds) = (validFrom.intersects(other.validFrom) || validTo.intersects(other.validTo))
+    fun overlaps(other: TemporalBounds) = overlapsValidTime(other) && overlapsSystemTime(other)
+
+    // note - this is not a symmetric operation
+
+    fun canBoundValidTime(other: TemporalBounds) =  overlapsValidTime(other) && validTo.lower <= other.validTo.upper
+    fun canBoundSytemTime(other: TemporalBounds) =  overlapsSystemTime(other) && systemTo.lower <= other.systemTo.upper
+    fun canBound(other: TemporalBounds) = canBoundValidTime(other) && canBoundSytemTime(other)
+
+    fun union(other: TemporalBounds) = TemporalBounds(
+        validFrom = validFrom.union(other.validFrom).union(other.validTo),
+        validTo = validTo.union(other.validFrom).union(other.validTo),
+        systemFrom = systemFrom.union(other.systemFrom),
+        systemTo = systemTo.union(other.systemTo))
```

Comment:
There is a lot of stuff here which is currently unused. I was just switching often between these and also wanted to include the symmetric operations. I can also slim this done if wished for.

Reply:
this suggests that we might have mis-modelled the valid-time and system-time parts of the TemporalBounds?

e.g. maybe rather than TB (vf,vt,sf,st) -> TC(lower, upper) we change it up slightly to TB(valid, sys) -> TC(min-from, max-from, min-from, min-to)?

Reply:
See https://github.com/xtdb/xtdb/pull/3583